timidity: add enableVorbis option

Niols opened the pull request:
###### Description of changes

TiMidity++ can not only play sounds, but also can save the generated waveforms as various audio file formats. The current package only supports Linux and Darwin outputs to play sound, but it does not support saving sounds as eg. vorbis.

This PR adds an optional flag `enableVorbis` to the package, disabled by default. This does not change the semantics of the current package but it allows users that need such support to curate their own set of packages with a `callPackage timidity { enableVorbis = true; }`.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Niols committed:
timidity: add enableVorbis option

TiMidity++ can not only play sounds, but also can save the generated
waveforms as various audio file formats. The current package only
supports Linux and Darwin outputs to play sound, but it does not support
saving sounds as eg. vorbis.

This commit adds an optional flag `enableVorbis` to the package,
disabled by default. This does not change the semantics of the current
package but it allows users that need such support to curate their own
set of packages with a `callPackage timidity { enableVorbis = true; }`.

turion commented:
Looking great! Are you motivated to add an integration test by any chance?

turion commented:
The corresponding section in the manual is https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests. Also, have a look into the directory `nixos/tests`, there are many examples one can learn from. You could even write a parametrized test that runs once with vorbis and once without.

Niols committed:
Add basic NixOS tests for TiMidity++

Niols commented:
@turion First time playing with NixOS tests, WDYT about d83ab7baf6d? I only really test the generation of Vorbis, though, because I have never used TiMidity++ for other uses like the ones the maintainer seems to have.

turion commented:
That's pretty cool!

@ofborg test timidity

turion commented:
Unfortunately, the test doesn't work on aarch64-linux. Is it architecture dependent (int sizes?) or is there maybe some nondeterministic setting like automatic bitrate detection?

turion commented:
(But the error message was very good!)

Niols commented:
Hum. It's not such a surprise to see that the bitrate and the size can be a bit flaky; it was maybe a bit dumb of me to rely on them for my test. Should I maybe just test the duration of the output file? That should be pretty stable.

Niols commented:
@ofborg test timidity

turion reviewed (changes requested):
Can you add a `passthru.tests` as described in https://nixos.org/manual/nixpkgs/stable/#ssec-nixos-tests-linking?

Also, why don't the tests run on darwin? I can't find any place forbidding this.

Niols committed:
Link TiMIDIty++ tests to the package

Niols commented:
I see I have been letting this PR wither for more than a year; no more! I added the passthru tests as per your request, I hope I did it correctly. I also rebased to a more recent version of `master`.

turion commented on nixos/tests/all-tests.nix:
```diff
@@ -1041,6 +1041,7 @@ in {
   tika = runTest ./tika.nix;
   timescaledb = handleTest ./timescaledb.nix {};
   timezone = handleTest ./timezone.nix {};
+  timidity = handleTest ./timidity {};
```
```suggestion
  timidity = handleTestOn ["aarch64-linux" "x86_64-linux"] ./timidity {};
```

Maybe `handleTestOn` to restrict to the systems where the test actually works?

Niols committed:
Update nixos/tests/all-tests.nix

Co-authored-by: Manuel Bärenz <programming@manuelbaerenz.de>

turion commented:
I find this fine to merge and would merge in a couple of days if noone objects.

Niols commented:
Yay :tada:

Niols commented:
Does no answer for two weeks mean that it is ready to be merged?

Niols commented:
Thank you very much, @turion, for your help on this over just a little less than two years!
